extensions/amp-mega-menu/0.1/amp-mega-menu.js

```diff
@@ -94,79 +114,99 @@ export class AmpMegaMenu extends AMP.BaseElement {
     this.registerMenuItems_();
     // items may not be present after build if dynamically rendered via amp-list,
     // in which case register them after DOM update instead.
-    this.element.addEventListener(AmpEvents.DOM_UPDATE, () => {
-      this.registerMenuItems_();
-    });
+    this.element.addEventListener(AmpEvents.DOM_UPDATE, this.domUpdateHandler_);
+
+    if (!this.maskElement_) {
+      const mask = this.document_.createElement('div');
+      mask.classList.add('i-amphtml-mega-menu-mask');
+      mask.setAttribute('aria-hidden', 'true');
+      // append mask to header so that all children of header appear above the mask
+      const maskParent =
+        closestAncestorElementBySelector(this.element, 'header') ||
+        this.element;
+      maskParent.classList.add('i-amphtml-mega-menu-mask-parent');
+      maskParent.appendChild(mask);
+      this.maskElement_ = mask;
+    }
 
-    const mask = this.document_.createElement('div');
-    mask.classList.add('i-amphtml-mega-menu-mask');
-    mask.setAttribute('aria-hidden', 'true');
-    // append mask to header so that all children of header appear above the mask
-    const maskParent =
-      closestAncestorElementBySelector(this.element, 'header') || this.element;
-    maskParent.classList.add('i-amphtml-mega-menu-mask-parent');
-    maskParent.appendChild(mask);
-    this.maskElement_ = mask;
-
-    this.documentElement_.addEventListener('click', () => this.collapse_());
-    this.documentElement_.addEventListener('keydown', event => {
-      // Collapse mega menu on ESC.
-      if (event.key === Keys.ESCAPE && this.collapse_()) {
-        event.preventDefault();
-      }
-    });
+    this.documentElement_.addEventListener('click', this.documentClickHandler_);
+    this.documentElement_.addEventListener(
+      'keydown',
+      this.documentKeyDownHandler_
+    );
 
     return Promise.resolve();
   }
 
   /** @override */
   unlayoutCallback() {
-    // TODO(#25047): unregister event listeners.
+    this.element.removeEventListener(
+      AmpEvents.DOM_UPDATE,
+      this.domUpdateHandler_
+    );
+    this.documentElement_.removeEventListener(
+      'click',
+      this.documentClickHandler_
+    );
+    this.documentElement_.removeEventListener(
+      'keydown',
+      this.documentKeyDownHandler_
+    );
+    this.unregisterMenuItemListeners_();
```

Comment:
Talked offline; will remove the unregister function and add/remove the event listeners for documentElement inside the expand/collapse methods since these listeners are only needed when a menu is open.